docs/proposals/testing.md

```diff
@@ -0,0 +1,129 @@
+*Note: we are focusing on a basic provision to start other actions should be added in the future.*
+
+### Design
+The base APB entry point will be able to find and run the test action. The test action will be a user defined playbook. 
+
+* To include the testing of an APB just add the playbook `test.yml`
+* The defaults for the test will be in the `vars/` directory of the playbooks.
+* The `verify_<name>` role should be in the roles folder. Should be a normal [ansible role](http://docs.ansible.com/ansible/latest/playbooks_reuse_roles.html).
+```bash
+my-apb/
+├── apb.yml
+├── Dockerfile
+├── playbooks/
+    ├── test.yaml  
+    └── vars/
+        └── test_defaults.yaml
+
+└── roles/
+    ├── ...
+    └── verify_apb_role
+        ├── defaults
+             └── defaults.yml
+        └── tasks  
+            └── main.yml
+```
+
+### Writing a `test.yaml` action
+To orchestrate the testing of an APB it is suggested to use the [include_vars](http://docs.ansible.com/ansible/latest/include_vars_module.html) and [include_role](http://docs.ansible.com/ansible/latest/include_role_module.html) modules.
+Example
+```yaml
+ - name: test rhscl-postgresql-apb
+   hosts: localhost
+   gather_facts: false
+   connection: local
+
+   # Load the ansible kubernetes modules
+   roles:
+   - role: ansible.kubernetes-modules
+     install_python_requirements: no
+ 
+   post_tasks:
+   # Include the default values needed for provision from test role.
+   - name: Load default variables for testing
+     include_vars: test_defaults.yaml
+   - name: Run the provisio role.
+     include_role:
+       name: rhscl-postgresql-apb-openshift
+   - name: Verify the provision.
+     include_role:
+       name: verify_rhscl-postgresql-apb-openshift
+```
+
+
+### Verify Roles
+Verify roles will allow the author to determine if the provision has failed or succeeded. 
+Example verify role.
+```yaml
+---
+ - name: url check for media wiki
+   uri:
+     url: "http://{{ route.route.spec.host }}"
+     return_content: yes
+   register: webpage
+   failed_when: webpage.status != 200
+```
+
+### Test Results
+The APB should be able to save test results so that an external caller can retrieve the results. This should behave very similar to [asb_encode_binding](https://github.com/fusor/ansible-asb-modules/blob/master/library/asb_encode_binding.py). 
+**Implementation changes to be made**: 
+
+- Create new module: `save_test_result` this will save the test results to `/var/tmp/test-result`. Should follow a format for the file such as 
+```text
+0
+success
+```
+or 
+```text
+1
+<message>
+```
+- Update [entrypoint.sh](https://github.com/fusor/apb-examples/blob/master/apb-base/files/usr/bin/entrypoint.sh) to wait with test-results were created.
+- Create `test-retrieval-init` to follow the same pattern as [bind-init](https://github.com/fusor/apb-examples/blob/master/apb-base/files/usr/bin/bind-init).
```

Comment:
I guess I'm not seeing why this has to be inside each APB. To me, it seems like each APB will go through the same test suite: Provision, if bindable; then bind; done. If that's the case, why not make a playbook/script/framework that runs externally to an APB and calls these actions? Why do we have to ship it with each APB?  And even if there are app specific validations/scripts in there, I think it would follow the same skeleton.

I think using an external framework is more pluggable.  It provides an imperative top down approach to testing, which makes it easier to create more complex tests and prevents us from having to carry test playbooks across all the example APBs.